lib/Dialect/ESI/ESIPasses.cpp

```diff
@@ -1302,6 +1304,278 @@ void ESItoHWPass::runOnOperation() {
+
+  // Classify each to_client req, appending non-local reqs to the list xor move
+  // into the block associated with the matching server.
+  SmallVector<RequestToClientConnection, 4> nonLocalToClientReqs;
+  mod.walk([&](RequestToClientConnection req) {
+    auto service = req.servicePortAttr().getModuleRef();
+    auto implOpF = localImplReqs.find(service);
+    if (implOpF == localImplReqs.end())
+      nonLocalToClientReqs.push_back(req);
+    else
+      req->moveBefore(implOpF->second, implOpF->second->end());
+  });
+
+  // Classify each to_server req, appending non-local reqs to the list xor move
+  // into the block associated with the matching server.
+  SmallVector<RequestToServerConnection, 4> nonLocalToServerReqs;
+  mod.walk([&](RequestToServerConnection req) {
+    auto service = req.servicePortAttr().getModuleRef();
+    auto implOpF = localImplReqs.find(service);
+    if (implOpF == localImplReqs.end())
+      nonLocalToServerReqs.push_back(req);
+    else
+      req->moveBefore(implOpF->second, implOpF->second->end());
+  });
+
+  // Replace each service instance with a generation request.
+  for (auto instOp :
+       llvm::make_early_inc_range(modBlock.getOps<ServiceInstanceOp>())) {
+    Block *portReqs = localImplReqs[instOp.service_symbolAttr()];
+    if (failed(replaceInst(instOp, portReqs)))
+      return failure();
+  }
+
+  // Surface all of the requests which cannot be fulfilled locally.
+  if (nonLocalToClientReqs.empty() && nonLocalToServerReqs.empty())
+    return success();
+  return surfaceReqs(mod, nonLocalToClientReqs, nonLocalToServerReqs);
+}
+
+LogicalResult ESIConnectServicesPass::replaceInst(ServiceInstanceOp instOp,
+                                                  Block *portReqs) {
+  assert(portReqs);
+
+  // Compute the result types for the new op -- the instance op's output types +
+  // the to_client types.
+  SmallVector<Type, 8> resultTypes(instOp.getResultTypes().begin(),
+                                   instOp.getResultTypes().end());
+  for (auto toClient : portReqs->getOps<RequestToClientConnection>())
+    resultTypes.push_back(toClient.receiving().getType());
+
+  // Compute the operands for the new op -- the instance op's operands + the
+  // to_server types. Reassign the reqs' operand to the new blocks arguments.
+  SmallVector<Value, 8> operands(instOp.getOperands().begin(),
+                                 instOp.getOperands().end());
+  for (auto toServer : portReqs->getOps<RequestToServerConnection>()) {
+    Value sending = toServer.sending();
+    operands.push_back(sending);
+    toServer.sendingMutable().assign(
+        portReqs->addArgument(sending.getType(), toServer.getLoc()));
+  }
+
+  // Create the generation request op.
+  OpBuilder b(instOp);
+  auto implOp = b.create<ServiceImplementReqOp>(
+      instOp.getLoc(), resultTypes, instOp.service_symbolAttr(),
+      instOp.identifierAttr(), operands);
+  implOp->setDialectAttrs(implOp->getDialectAttrs());
```

Comment:
why?

Reply:
Was this meant to be `implOp->setDialectAttrs(instOp->getDialectAttrs())`

Reply:
yep